src/Storages/MergeTree/MergeTreeIndexReadResultPool.cpp

```diff
@@ -94,13 +101,346 @@ SkipIndexReadResultPtr MergeTreeSkipIndexReader::read(const RangesInDataPart & p
+        while (val < ending_row)
+        {
+            has_value = true;
+            pos[val - starting_row] = 1;
+            if (!roaring::api::roaring64_iterator_advance(it))
+                break;
+            val = roaring64_iterator_value(it);
+        }
+        roaring::api::roaring64_iterator_free(it);
+        return has_value;
+    }
+}
+
+template ProjectionIndexBitmapPtr ProjectionIndexBitmap::createFromRange<UInt32>(UInt32 start, UInt32 end);
+template ProjectionIndexBitmapPtr ProjectionIndexBitmap::createFromRange<UInt64>(UInt64 start, UInt64 end);
+
+template bool ProjectionIndexBitmap::contains<UInt32>(UInt32 value);
+template bool ProjectionIndexBitmap::contains<UInt64>(UInt64 value);
+
+template void ProjectionIndexBitmap::add<UInt32>(UInt32 value);
+template void ProjectionIndexBitmap::add<UInt64>(UInt64 value);
+
+SingleProjectionIndexReader::SingleProjectionIndexReader(
+    std::shared_ptr<MergeTreeReadPoolProjectionIndex> pool,
+    PrewhereInfoPtr prewhere_info,
+    const ExpressionActionsSettings & actions_settings,
+    const MergeTreeReaderSettings & reader_settings)
+    : projection_index_read_pool(std::move(pool))
+    , processor(std::make_unique<MergeTreeSelectProcessor>(
+          std::static_pointer_cast<IMergeTreeReadPool>(projection_index_read_pool),
+          std::make_unique<MergeTreeProjectionIndexSelectAlgorithm>(),
+          std::move(prewhere_info),
+          nullptr /*lazily_read_info*/,
+          actions_settings,
+          reader_settings))
+{
+}
+
+ProjectionIndexBitmapPtr SingleProjectionIndexReader::read(const RangesInDataPart & ranges)
+{
+    bool can_use_32bit_part_offset = ranges.parent_ranges.max_part_offset <= std::numeric_limits<UInt32>::max();
+
+    /// Prepare the read processor with the current part and its read ranges.
+    /// This sets up internal state needed to read the projection data.
+    assert_cast<MergeTreeProjectionIndexSelectAlgorithm &>(*processor->algorithm).preparePartToRead(&ranges);
+    auto res = can_use_32bit_part_offset ? ProjectionIndexBitmap::create32() : ProjectionIndexBitmap::create64();
+
+    /// Start reading chunks from the projection index reader.
+    /// Each chunk contains a column of UInt64 offsets that we insert into the bitmap.
+    while (true)
+    {
+        auto chunk = processor->read();
+        if (chunk.chunk)
+        {
+            if (chunk.chunk.getNumRows() > 0)
+            {
+                chassert(chunk.chunk.getColumns().size() == 1);
+                auto offset_column = chunk.chunk.getColumns()[0]->convertToFullIfNeeded();
+                const auto & offsets = assert_cast<const ColumnUInt64 &>(*offset_column);
+
+                auto add_offsets = [&]<typename Offset>(Offset)
+                {
+                    if (ranges.parent_ranges.isContiguousFullRange())
+                    {
+                        for (auto offset : offsets.getData())
+                            res->add<Offset>(offset);
```

Comment:
Did you consider using `roaring_bitmap_add_bulk`? According to the [documentation](https://github.com/RoaringBitmap/CRoaring/blob/9b7cc0ff1c41e9457efb6228cfd2c538d0155303/include/roaring/roaring.h#L300-L301), it can improve performance if consecutive values have equal highest 16 bits. I don't know whether it's a typical situation, but it looks like this optimization doesn't give a significant overhead when cannot be applied.